src/session/client-certificate.c

```diff
@@ -127,6 +132,97 @@ unsigned long long get_proc_pid_start_time (int dirfd)
+      ucred_len < sizeof ucred)
+    {
+      debug ("failed to read stdin peer credentials: %m; not in socket mode?");
+      warnx ("Certificate authentication only supported with cockpit-session.socket");
+      exit_init_problem ("authentication-unavailable", "Certificate authentication only supported with cockpit-session.socket");
+    }
+
+  debug ("unix socket mode, ws peer pid %d", ucred.pid);
+  int ws_pid_dirfd = open_proc_pid (ucred.pid);
+  unsigned long long ws_start_time = get_proc_pid_start_time (ws_pid_dirfd);
+
+  int my_pid_dirfd = open_proc_pid (getpid ());
+  unsigned long long my_start_time = get_proc_pid_start_time (my_pid_dirfd);
+  close (my_pid_dirfd);
+
+  debug ("peer start time: %llu, my start time: %llu", ws_start_time, my_start_time);
+
+  /* Guard against pid recycling: If a malicious user captures ws, keeps the socket in a forked child and exits
+    * the original pid, they can trap a different user to login, get the old pid (pointing ot their cgroup), and
+    * capture their session. To prevent that, require that ws must have started earlier than ourselves. */
+  if (my_start_time < ws_start_time)
+    {
+      warnx ("start time of this process (%llu) is older than cockpit-ws (%llu), pid recycling attack?",
+              my_start_time, ws_start_time);
+      close (ws_pid_dirfd);
+      exit_init_problem ("access-denied", "implausible cockpit-ws start time");
+    }
+
+  return ws_pid_dirfd;
+}
+
+/* Get a /proc/[pid] dirfd for our Unix socket peer (i.e. cockpit-ws).
+ * We only support being called via cockpit-session.socket (i.e. Unix socket).
+ */
+static int
+get_ws_proc_fd (void)
+{
+#if defined(SO_PEERPIDFD) && defined(HAVE_PIDFD_GETPID)
+  int pidfd = -1;
+  socklen_t socklen = sizeof pidfd;
+  /* this is always the pidfd for the process that started the communication, it cannot be recycled */
+  if (getsockopt (STDIN_FILENO, SOL_SOCKET, SO_PEERPIDFD, &pidfd, &socklen) < 0)
+    {
+      if (errno == ENOPROTOOPT || errno == ENODATA)
+        {
+          debug ("SO_PEERPIDFD not supported: %m, falling back to pid/time check");
+          return get_ws_proc_fd_pid_time ();
+        }
+
+      warn ("Failed to get peer pidfd");
+      exit_init_problem ("access-denied", "Failed to get peer pidfd");
+    }
+  /* this is an inout parameter, be extra suspicious; this really Should Not Happen™, so bomb out */
+  if (socklen < sizeof pidfd)
+    errx (EX, "SO_PEERPIDFD returned too small result");
+
+  /* get pid for pidfd; from here on this is racy and could suffer from PID recycling */
+  pid_t pid = pidfd_getpid (pidfd);
+  if (pid < 0)
+    {
+      warn ("Failed to get pid from pidfd");
+      exit_init_problem ("access-denied", "Failed to get pid from pidfd");
+    }
+
+  debug ("pid from ws peer pidfd: %i", (int) pid);
+  int ws_pid_dirfd = open_proc_pid (pid);
+
+  /* check that the pid is still valid to guard against recycling */
+  if (pidfd_getpid (pidfd) != pid)
```

Comment:
This screams of an ABA, but I think it's safe, right?  Because we'd never see the second "A", but rather `ESRCH`.

Reply:
Not sure what "ABA" is, but indeed this catches both "pid changed" (which shouldn't really happen) and "query failed", i.e. -1. We already asserted that pid is positive.

Reply:
"It's kinda like ABBA, but with 50% less B".

But seriously:

https://en.wikipedia.org/wiki/ABA_problem

It's this pattern of "operation X can only be done safely under condition Y = A", so you:
 - check Y == A
 - do X
 - recheck Y == A

and assume that condition Y == A held the whole time, but actually it changed to B while you were doing X and then changed back to A after (hence ABA).  It's not possible in this case, though, since there is only one state transition possible, and it's permanent: a running PID becomes `ESCHR` via the process exiting.